File: python/socketd/socketd/transport/CodecByteBuffer.py
from typing import Callable
from .Codec import Codec
from socketd.core.module.Frame import Frame
from socketd.core.module.MessageDefault import MessageDefault
from socketd.core.module.EntityDefault import EntityDefault
from socketd.core.Costants import Flag, Constants
from socketd.core.config.Config import Config
from ..core.Buffer import Buffer

import os


def assert_size(name: str, size: int, limitSize: int) -> None:
    if size > limitSize:
        buf = f"This message {name} size is out of limit {limitSize} ({size})"
        raise RuntimeError(buf)


class CodecByteBuffer(Codec):
    def __init__(self, config: Config):
        self.config = config

    def write(self, frame: Frame, factory: Callable[[int], Buffer]) -> Buffer:
        if frame.message is None:
            # length (flag + int.bytes)
            _len = 2 * 4
            target = factory(_len)

            # length
            target.put_int(_len)
            # flag
            target.put_int(frame.flag)
            target.flush()

            return target
        else:
            # sid
            sidB: bytes = frame.message.get_sid().encode(self.config.get_charset())
            # event
            event: bytes = frame.message.get_event().encode(self.config.get_charset())
            # metaString
            metaStringB: bytes = frame.message.get_entity().get_meta_string().encode(self.config.get_charset())

            # length (flag + sid + event + metaString + data + int.bytes + \n*3)
            len1 = len(sidB) + len(event) + len(
                metaStringB) + frame.message.get_entity().get_data_size() + 1 * 3 + 2 * 4

            assert_size("sid", len(sidB), Constants.MAX_SIZE_SID)
            assert_size("event", len(event), Constants.MAX_SIZE_EVENT)
            assert_size("metaString", len(metaStringB), Constants.MAX_SIZE_META_STRING)
            assert_size("data", frame.message.get_entity().get_data_size(), Constants.MAX_SIZE_FRAGMENT)

            target: Buffer = factory(len1)

            # length
            target.put_int(len1)

            # flag
            target.put_int(frame.flag)

            # sid
            target.write(sidB)
            target.write(b'\n')

            # event
            target.write(event)
            target.write(b'\n')

            # metaString
            target.write(metaStringB)
            target.write(b'\n')

            # data
            if frame.message.get_entity().get_data() is not None:
                target.write(frame.message.get_entity().get_data().getvalue())
            target.flush()

            return target

    def read(self, buffer: Buffer) -> Frame:
        len0 = buffer.get_int()

        if len0 > (buffer.remaining() + 4):
            return None

        flag = buffer.get_int()  # 取前一位数据

        if len0 == 8:
            # len + flag
            return Frame(Flag.of(flag), None)
        else:
            metaBufSize = min(Constants.MAX_SIZE_META_STRING, buffer.remaining())
            # 1. decode sid and event
            by = Buffer(limit=metaBufSize)
            sid = self.decodeString(buffer, by, Constants.MAX_SIZE_SID)
            event = self.decodeString(buffer, by, Constants.MAX_SIZE_EVENT)
            metaString = self.decodeString(buffer, by, Constants.MAX_SIZE_META_STRING)

            # 2. decode body
            dataRealSize = len0 - buffer.tell()
            data: bytearray = None
            if dataRealSize > Constants.MAX_SIZE_FRAGMENT:
                # exceeded the limit, read and discard the bytes
                data = bytearray(Constants.MAX_SIZE_FRAGMENT)
                buffer.readinto(data)
                for i in range(dataRealSize - Constants.MAX_SIZE_FRAGMENT):
                    buffer.read()
            else:
                data = bytearray(buffer.read(dataRealSize))

            message = MessageDefault().set_sid(sid).set_event(event).set_entity(
                EntityDefault().set_meta_string(metaString).set_data(data)
            )
            message.flag = Flag.of(flag)
            return Frame(message.flag, message)

    def decodeString(self, reader: Buffer, buf: Buffer, maxLen: int) -> str:
        b = bytearray(reader.readline(maxLen).replace(b'\n', b''))
        if buf.limit() < 1:
            return ""
        return b.decode(self.config.get_charset())

File: javascript/socketd-examples/demo01-js-ws-npm/js/socketd/transport/core/CodecByteBuffer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CodecByteBuffer = void 0;
const Frame_1 = require("./Frame");
const StrUtils_1 = require("../../utils/StrUtils");
const Asserts_1 = require("./Asserts");
const Constants_1 = require("./Constants");
const Message_1 = require("./Message");
const Entity_1 = require("./Entity");
/**
 * 编解码器（基于 CodecReader,CodecWriter 接口编解）
 *
 * @author noear
 * @since 2.0
 */
class CodecByteBuffer {
    constructor(config) {
        this._config = config;
    }
    /**
     * 解码写入
     *
     * @param frame         帧
     * @param targetFactory 目标工厂
     */
    write(frame, targetFactory) {
        if (frame.message()) {
            //sid
            const sidB = StrUtils_1.StrUtils.strToBuf(frame.message().sid(), this._config.getCharset());
            //event
            const eventB = StrUtils_1.StrUtils.strToBuf(frame.message().event(), this._config.getCharset());
            //metaString
            const metaStringB = StrUtils_1.StrUtils.strToBuf(frame.message().metaString(), this._config.getCharset());
            //length (len[int] + flag[int] + sid + event + metaString + data + \n*3)
            const frameSize = 4 + 4 + sidB.byteLength + eventB.byteLength + metaStringB.byteLength + frame.message().dataSize() + 2 * 3;
            Asserts_1.Asserts.assertSize("sid", sidB.byteLength, Constants_1.Constants.MAX_SIZE_SID);
            Asserts_1.Asserts.assertSize("event", eventB.byteLength, Constants_1.Constants.MAX_SIZE_EVENT);
            Asserts_1.Asserts.assertSize("metaString", metaStringB.byteLength, Constants_1.Constants.MAX_SIZE_META_STRING);
            Asserts_1.Asserts.assertSize("data", frame.message().dataSize(), Constants_1.Constants.MAX_SIZE_DATA);
            const target = targetFactory(frameSize);
            //长度
            target.putInt(frameSize);
            //flag
            target.putInt(frame.flag());
            //sid
            target.putBytes(sidB);
            target.putChar('\n'.charCodeAt(0));
            //event
            target.putBytes(eventB);
            target.putChar('\n'.charCodeAt(0));
            //metaString
            target.putBytes(metaStringB);
            target.putChar('\n'.charCodeAt(0));
            //data
            target.putBytes(frame.message().data().getArray());
            target.flush();
            return target;
        }
        else {
            //length (len[int] + flag[int])
            const frameSize = 4 + 4;
            const target = targetFactory(frameSize);
            //长度
            target.putInt(frameSize);
            //flag
            target.putInt(frame.flag());
            target.flush();
            return target;
        }
    }
    /**
     * 编码读取
     *
     * @param buffer 缓冲
     */
    read(buffer) {
        const frameSize = buffer.getInt();
        if (frameSize > (buffer.remaining() + 4)) {
            return null;
        }
        const flag = buffer.getInt();
        if (frameSize == 8) {
            //len[int] + flag[int]
            return new Frame_1.Frame(Constants_1.Flags.of(flag), null);
        }
        else {
            const metaBufSize = Math.min(Constants_1.Constants.MAX_SIZE_META_STRING, buffer.remaining());
            //1.解码 sid and event
            const buf = new ArrayBuffer(metaBufSize);
            //sid
            const sid = this.decodeString(buffer, buf, Constants_1.Constants.MAX_SIZE_SID);
            //event
            const event = this.decodeString(buffer, buf, Constants_1.Constants.MAX_SIZE_EVENT);
            //metaString
            const metaString = this.decodeString(buffer, buf, Constants_1.Constants.MAX_SIZE_META_STRING);
            //2.解码 body
            const dataRealSize = frameSize - buffer.position();
            let data;
            if (dataRealSize > Constants_1.Constants.MAX_SIZE_DATA) {
                //超界了，空读。必须读，不然协议流会坏掉
                data = new ArrayBuffer(Constants_1.Constants.MAX_SIZE_DATA);
                buffer.getBytes(data, 0, Constants_1.Constants.MAX_SIZE_DATA);
                for (let i = dataRealSize - Constants_1.Constants.MAX_SIZE_DATA; i > 0; i--) {
                    buffer.getByte();
                }
            }
            else {
                data = new ArrayBuffer(dataRealSize);
                if (dataRealSize > 0) {
                    buffer.getBytes(data, 0, dataRealSize);
                }
            }
            //先 data , 后 metaString (避免 data 时修改元信息)
            const message = new Message_1.MessageBuilder()
                .flag(Constants_1.Flags.of(flag))
                .sid(sid)
                .event(event)
                .entity(new Entity_1.EntityDefault().dataSet(data).metaStringSet(metaString))
                .build();
            return new Frame_1.Frame(message.flag(), message);
        }
    }
    decodeString(reader, buf, maxLen) {
        const bufView = new DataView(buf);
        let bufViewIdx = 0;
        while (true) {
            const c = reader.getByte();
            if (c == 10) { //10:'\n'
                break;
            }
            if (maxLen > 0 && maxLen <= bufViewIdx) {
                //超界了，空读。必须读，不然协议流会坏掉
            }
            else {
                if (c != 0) { //32:' '
                    bufView.setInt8(bufViewIdx, c);
                    bufViewIdx++;
                }
            }
        }
        if (bufViewIdx < 1) {
            return "";
        }
        //这里要加个长度控制
        return StrUtils_1.StrUtils.bufToStr(buf, 0, bufViewIdx, this._config.getCharset());
    }
}
exports.CodecByteBuffer = CodecByteBuffer;
